### ISSUE TITLE
Performance statistics report

### ISSUE THREAD

**Issue by jeromekelleher** [tskit-dev/tskit#2444]
It would be very useful to have some simple microbenchmarks we could run from time-to-time. We've tried using ASV, but it's very heavyweight and collects masses of data that's never used. 

Requirements:

- Make a script ``perf_benchmark.py`` which runs a standard set of microbenchmarks and outputs the result to a file. The idea is this file should be updated before every release, so that we can spot any perf regressions and also so that we have a per-release history of the benchmarks in the history.
- Update the developer docs to include how and when to run these benchmarks as part of the release process.

CPU time performance benchmarks:

Given a standard file

- [x] time to load
- [x] time to save
- [x] time to access ``ts.tables`` in a loop
- [x] time to access ``tables.nodes``, ``tables.individuals`` etc
- [x] time to access columns, ``nodes.flags``, etc
- [x] time to get first tree, ts.first()
- [x] time to seek to middle tree
- [x] time to iterate over all trees
- [x] time to access tree arrays, tree.parent etc.
- [x] time to decode all variants
- [x] time to iterate over all rows, with and without metadata
- [x] time to write to vcf (writing to devnull)

additional as I thought of them (BJ):
- [x] tree node arrays (`postorder` et al)
- [x] tree row accessors (`ts.node(42)`)
- [x] tree accessors (`tree.left_sib(42)`)
- [x] iterate over `tree.nodes`

It would be nice to track the memory usage here, but this may be much more difficult and not worth the effort.

**Comment by benjeffery** [comment]
Done some research here, measuring the variance of different methods. The main aim is to avoid this phenomenon in micro benchmarks, where running code over a period of 20s reveals two performance regimes:
![no_running](https://user-images.githubusercontent.com/8552/181750485-32c40cc8-2097-4068-bffc-5fb2d808a781.png)

Using `timeit` and having a balance between "number of trials in the tight loop" and "number of repetitions of the loop" seems to avoid this phenomena and gives lowest variance. I've converged on:
```python
def autotime(setup: str, code: str)
    t = timeit.Timer(setup=setup, stmt=code)
    # We want each test to last around 1-2 seconds, and to have a good number of trials and
    # repetitions as this gives the lowest variance in testing.
    one_run = t.timeit(number=1)
    num_trials = int(max(1, 0.01 / one_run))
    num_repeats = math.ceil(2 / (one_run * num_trials))
    t1 = time.time()
    results = t.repeat(repeat=num_repeats, number=num_trials)
    return np.mean(results) / num_trials
```

For configuration this is the proposed format:
```yaml
setup: |
  import sys
  import tskit
  import msprime
  import time
  ts = msprime.simulate(10, random_seed=1)

benchmarks:
  - code: ts.tables
  - code: tables.{table}
    setup: tables = ts.tables
    parameters:
      table:
        - nodes
        - edges
        - sites
        - mutations
        - populations
        - individuals
        - migrations
        - provenances
  - code: "{table}.{column}"
    setup: |
      tables = ts.tables
      {table} = tables.{table}
    parameters:
      table:
        nodes:
          column:
            - flags
        edges:
          column:
            - left
            - right
            - parent
            - child
....
```
This allows parameterisation in both the setup and code.

WIP at https://github.com/tskit-dev/tskit/compare/main...benjeffery:bench

**Comment by jeromekelleher** [comment]
Very nice, I like it!

**Comment by benjeffery** [comment]
Tests with side effects are tricky as `timeit` does setup once, then repeats them. So for example `seek` only does meaningful work on the first iteration of the loop. For `seek` I've fixed this by seeking backwards and forwards in the loop. If I come across one where such a trick is not possible we'll just have to only have one trial per loop.

**Comment by benjeffery** [comment]
Closed by #2454
